beacon_node/store/src/lib.rs

```diff
@@ -58,8 +51,13 @@ pub trait Store<E: EthSpec>: Sync + Send + Sized + 'static {
     /// Removes `key` from `column`.
     fn key_delete(&self, column: &str, key: &[u8]) -> Result<(), Error>;
 
+    /// Execute either all of the operations in `batch` or none at all, returning an error.
+    fn do_atomically(&self, batch: &[StoreOp]) -> Result<(), Error>;
```

Comment:
I'm torn about `do_atomically`. On one hand it seems categorically different from the other `KeyValueStore` methods in that it deals with high-level objects (states, blocks, etc) rather than bytes. On the other hand, any atomicity implementation must be specific to the underlying KV store. I think we could get away with removing it for now (and implementing it at the `Store` level instead), but it might clash with my other idea of making `HotColdDB` generic over its key value store (see main comment).

Reply:
Future PR.

Reply:
Ultimately, `do_atomically` should exist at two levels: At the `KeyValueStore` level, where it works with raw DB ops like PutKeyValue, DeleteKey, etc.  And at the `Store` level, where it works with high level ops: PutBlock, DeleteState, etc.  The high level one translates operations into low level ones and and calls `KeyValueStoere::do_atomically` on them.